### PR TITLE
[tune] clean up population based training prototype

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

- elide checkpointing overhead when between perturbation intervals
- clean up constructor params
- better status output in console
- add example to tune docs

PBT discovering better hyperparams:
train

### CODE REVIEW

File: python/ray/tune/pbt.py [hunk 1]
```diff
@@ -2,189 +2,248 @@
...
+        return str((
+            self.last_score, self.last_checkpoint,
+            self.last_perturbation_time))
+
+
+def explore(config, mutations, resample_probability):
+    """Return a config perturbed as specified.
+
+    Args:
+        config (dict) Original hyperparameter configuration.
```
Comment (richardliaw) [inline python/ray/tune/pbt.py]: are colons needed here to show up on readthedocs? (`config (dict): `)

File: python/ray/tune/pbt.py [hunk 2]
```diff
@@ -2,189 +2,248 @@
...
+        mutations (dict) Specification of mutations to perform as documented
+            in the PopulationBasedTraining scheduler.
+        resample_probability (bool) Probability of allowing resampling of a
```
Comment (richardliaw) [inline python/ray/tune/pbt.py]: float, not bool, right?

File: python/ray/tune/pbt.py [hunk 3]
```diff
@@ -2,189 +2,248 @@
...
-            compare model to other models.
-        perturbation_interval (float): Used in the truncation ready function to
-            determine if enough time has passed so that a agent can be tested
-            for readiness.
-        hyperparameter_mutations (dict); Possible values that each
-            hyperparameter can mutate to, as certain hyperparameters
-            only work with certain values.
+            with `time_attr`, this may refer to any objective value. Stopping
+            procedures will use this attribute.
+        perturbation_interval (float): Models will be considered for
```
Comment (richardliaw) [inline python/ray/tune/pbt.py]: can you explicitly clarify here that this interval is in units of `time_attr`? It was pretty unclear, especially given the example being labeled "10 iterations"

File: python/ray/tune/pbt.py [hunk 4]
```diff
@@ -2,189 +2,248 @@
...
+        self._num_perturbations += 1
+
+    def _quantiles(self):
+        """Returns trials in the lower and upper `quantile` of the population.
+
+        If there is not enough data to compute this, returns empty lists."""
+
+        trials = []
+        for trial, state in self._trial_state.items():
+            if state.last_score is not None and not trial.is_finished():
```
Comment (richardliaw) [inline python/ray/tune/pbt.py]: why not just return ranges? sort the last_score and then take the quartile points.

File: python/ray/tune/pbt.py [hunk 5]
```diff
@@ -2,189 +2,248 @@
...
-                return TrialScheduler.CONTINUE
-            else:
-                self._exploit(self._hyperparameter_mutations, best_trial,
-                              trial, trial_runner, time)
-                return TrialScheduler.CONTINUE
+            state.last_checkpoint = None  # not a top trial
+
+        if trial in lower_quantile:
+            trial_to_clone = random.choice(upper_quantile)
+            assert trial != trial_to_clone
```
Comment (richardliaw) [inline python/ray/tune/pbt.py]: `is not` would probably be better

**Review by richardliaw — COMMENTED** [review]
this is a first read-through; will test out tomorrow

File: python/ray/tune/examples/pbt_example.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import json
+import os
+import random
+import time
+
+import numpy as np
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/examples/pbt_example.py]: Switched to just use random

File: python/ray/tune/examples/pbt_example.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    def _setup(self):
+        self.timestep = 0
+        self.current_value = 0.0
+
+    def _train(self):
+        time.sleep(0.1)
+
+        # Reward increase is parabolic as a function of factor_2, with a
+        # maximia around factor_1=10.0.
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/examples/pbt_example.py]: Fixed

File: python/ray/tune/pbt.py [hunk 8]
```diff
@@ -2,189 +2,248 @@
 from __future__ import division
 from __future__ import print_function
 
-import collections
 import random
 import math
 import copy
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/pbt.py]: I prefer to use the built-in libraries when possible

File: python/ray/tune/pbt.py (hunk 1 above)
Comment (ericl) [reply python/ray/tune/pbt.py]: Fixed

File: python/ray/tune/pbt.py (hunk 2 above)
Comment (ericl) [reply python/ray/tune/pbt.py]: Fixed

File: python/ray/tune/pbt.py [hunk 9]
```diff
@@ -2,189 +2,248 @@
...
 class PopulationBasedTraining(FIFOScheduler):
-    """Implements the Population Based Training algorithm as described in the
-    PBT paper (https://arxiv.org/abs/1711.09846)(Experimental):
+    """Implements the Population Based Training (PBT) algorithm.
+
+    https://deepmind.com/blog/population-based-training-neural-networks
+
+    PBT trains a group of models (or agents) in parallel. Periodically, poorly
+    performing models clone the state of the top performers, and a random
+    mutation is applied to their hyper-parameters in the hopes of
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/pbt.py]: Fixed

File: python/ray/tune/pbt.py (hunk 3 above)
Comment (ericl) [reply python/ray/tune/pbt.py]: Fixed

File: python/ray/tune/pbt.py (hunk 5 above)
Comment (ericl) [reply python/ray/tune/pbt.py]: Done

File: python/ray/tune/pbt.py (hunk 4 above)
Comment (ericl) [reply python/ray/tune/pbt.py]: Set semantics are slightly less error-prone, and here the performance doesn't matter so much so it's fine to take the hit

File: python/ray/tune/tune.py [hunk 10]
```diff
@@ -44,6 +44,7 @@ def run_experiments(experiments, scheduler=None, with_server=False,
 
     for name, spec in experiments.items():
         for trial in generate_trials(spec, name):
+            trial.verbose = verbose
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/tune.py]: Fixed

**Comment by ericl** [comment]
Comments updated

File: python/ray/tune/pbt.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (robertnishihara) [reply python/ray/tune/pbt.py]: Given that we're already using `numpy` so heavily, wouldn't it make sense to just use `numpy` and minimize the number of libraries that we use?

[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/pbt.py]: They're part of the python runtime already, so there is no extra cost here.

[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tune/pbt.py]: I would propose the following
- Whenever possible, use the built-in libraries
- Use numpy/other libraries only when the built-in libraries do not satisfy your requirement

That way, we are always using the simplest possible function call that works.

**Review by richardliaw — COMMENTED** [review]
(quick review - still in progress)

**Review by richardliaw — APPROVED** [review]
Seems functional on my computer - performance on cartpole doesn't seem "amazing".

**Pull request merged**